### PR TITLE
470 explain tgw

### LINKED ISSUE #470: explainability: support tgw

### PR DESCRIPTION
**Pull request by ShiriMoran**
fixes #470

### CODE REVIEW

File: pkg/ibmvpc/vpc.go [hunk 1]
```diff
@@ -753,6 +780,111 @@ func (tgw *TransitGateway) AllowedConnectivity(src, dst vpcmodel.VPCResourceIntf
...
+	return nil
+}
+
+func (fip *FloatingIP) StringPrefixDetails(src, dst vpcmodel.Node, verbose bool) (string, error) {
+	return "", nil
+}
+
+func (pgw *PublicGateway) StringPrefixDetails(src, dst vpcmodel.Node, verbose bool) (string, error) {
+	return "", nil
+}
```
Comment (haim-kermany) [inline pkg/ibmvpc/vpc.go:886]: put these  two methods in the right place

**Review by haim-kermany — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/parser.go [hunk 2]
```diff
@@ -794,6 +796,11 @@ func getTgwObjects(c *datamodel.ResourcesContainerModel,
 			// TGW's destSubnets contains subnets from its connected VPCs which are contained within routes from its table
 			tgwMap[tgwUID].destSubnets = append(tgwMap[tgwUID].destSubnets, getVPCdestSubnetsByAdvertisedRoutes(tgwMap[tgwUID], vpc)...)
 			tgwMap[tgwUID].addSourceAndDestNodes()
+
+			// explainability related computation
+			// vpcApsPrefixes is a map from the vpc's ap to the index of the prefix that matches it, -1 for default if no match
+			vpcApsPrefixes, _ := getVpcApsPrefixes(tgwConn, vpc) // if getVPCAdvertisedRoutes completed without an error, so would getVpcApsPrefixes
+			maps.Copy(tgwMap[tgwUID].vpcApsPrefixes, map[string]map[*ipblock.IPBlock]tgwPrefix{vpcUID: vpcApsPrefixes})
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/parser.go:841]: You are right, the copy is redundant. Fixed.

File: pkg/vpcmodel/explainabilityConnectivity.go [hunk 3]
```diff
@@ -30,15 +30,16 @@ type srcDstDetails struct {
 	egressEnabled  bool
 	// the connection between src to dst, in case the connection was not part of the query;
 	// the part of the connection relevant to the query otherwise.
-	conn   *connection.Set
-	router RoutingResource // the router (fip or pgw) to external network; nil if none
+	conn           *connection.Set
+	externalRouter RoutingResource // the router (fip or pgw) to external network; nil if none or not relevant
+	tgwRouter      RoutingResource // the tgw router between src and dst from different VPCs; nil if none or not relevant
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityConnectivity.go:35]: 1. Renamed to crossVpcRouter
2. Good question. Is it possible that connection between two VPCs is enabled by more than one tgw Router or by more than one tgw Connection in the same Router? @adisos @kyorav 
Currently the implementation assumes one of each (If indeed more than one router/connection is possible I would defer it to a different PR, perhaps even to V0.5 and for now just add a comment).

File: pkg/ibmvpc/vpc.go (hunk 1 above)
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go]: You mean with the other class functions in the same file? Done
Did you mean something else? what?

File: pkg/vpcmodel/explainabilityInput.go [hunk 4]
```diff
@@ -148,20 +150,39 @@ func noMatchErr(srcFoundSomeCfg, dstFoundSomeCfg bool, errMsgInternalNoConnected
...
+			} else {
+				moreThanOneSingleVpc = true
+			}
+		}
+	}
+	if singleVpcMatched == "" || moreThanOneSingleVpc {
+		matchConfigs := make([]string, len(configsWithSrcDstNode))
+		i := 0
+		for cfgID := range configsWithSrcDstNode {
+			matchConfigs[i] = configsMap[cfgID].VPC.Name()
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go:174]: No, this is after I already know there is a problem. I'm gathering the names of the relevant configs for the error message. A clarification comment added

File: pkg/vpcmodel/explainabilityInput.go [hunk 5]
```diff
@@ -119,7 +121,7 @@ func (configsMap MultipleVPCConfigs) getVPCConfigAndSrcDstNodes(src, dst string)
 		return noMatchErr(srcFoundSomeCfg, dstFoundSomeCfg, errMsgInternalNoConnectedVSI, errMsgInternalNotWithinSubnet,
 			errMsgNoValidSrc, errMsgNoValidDst)
 	default: // len(configsWithSrcDstNode) > 1: src and dst found in more than one VPC configs - error
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go:123]: right. fixed

File: pkg/vpcmodel/explainabilityInput.go [hunk 6]
```diff
@@ -119,7 +121,7 @@ func (configsMap MultipleVPCConfigs) getVPCConfigAndSrcDstNodes(src, dst string)
...
-		return configsMap.matchMoreThanOneCfgErr(src, dst, configsWithSrcDstNode)
+		return configsMap.matchMoreThanOneCfg(src, dst, configsWithSrcDstNode)
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: Excellent comment.
Done

File: pkg/vpcmodel/explainabilityInput.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/explainabilityInput.go]: do you want to add the name of the dummy vpc to the message?

File: pkg/vpcmodel/explainabilityConnectivity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kyorav) [reply pkg/vpcmodel/explainabilityConnectivity.go]: I don't think you can connect two TGSs sequentially, but you can connect a VPC to two TGWs. So if two VPCs are both connected to the same two TGWs there are two possible routes connecting them. In this situation IBM does not guarantee the behavior, i.e. you don't know which route would be taken and this may change arbitrarily. I don't know if there is an actual use-case for this. It is probably OK to raise a warning, or even an error saying we don't support this scenario.

File: pkg/vpcmodel/explainabilityInput.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/explainabilityInput.go]: still, you check the lens in two functions...

File: pkg/vpcmodel/explainabilityConnectivity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityConnectivity.go]: We already have an error in this case - since it will result in the `<src, dst>` being found in more than one `multiVpc` config. So in this stage I'll just clarify the error message and add that this scenario is not supported (yet).

File: pkg/vpcmodel/explainabilityInput.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: no. The dummy vpc is not relevant if the src and dst are of the same vpc

File: pkg/vpcmodel/explainabilityInput.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/vpcmodel/explainabilityInput.go]: @ShiriMoran , If you truly believe it makes sense to put all code in one huge, messy, deeply-nested function, you can always add a `// nolint` comment. I acknowledge there are (rare) cases when this is the best compromise. I'm certain PR reviewers will double-check that.

[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: No, this is not what I believe @nevo ;-) The fine tuning I had in mind was not to remove this lint error, but perhaps to change the number in which the error is asserted. I'll try to add this comment and restore the code @haim-kermany say what you think after I do

[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: @zivnevo @haim-kermany please look at the last commit and see if it is better, as I think it is.

File: pkg/vpcmodel/explainabilityConnectivity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityConnectivity.go]: done

File: pkg/vpcmodel/explainabilityInput.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/vpcmodel/explainabilityInput.go]: I think it's now better than when it was split into two functions.
I still find it somewhat hard to follow, and tend to agree with the linter 😞

[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/explainabilityInput.go]: it is OK to put it in two function.
1. you can put all the len checks in the second function
2. you can put all the len checks in the first function, and the error handling in the second

[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/explainabilityInput.go]: I think you need to move the switch of the error to another function

File: pkg/vpcmodel/explainabilityInput.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/explainabilityInput.go]: if I understand correctly, the config name of the dummy config is in the message.
however, these are not real configs, and the name are not valid to the user

File: pkg/vpcmodel/explainabilityInput.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: There are 4 strings + and error that tiny function will need to return. Not sure it makes sense.

File: pkg/vpcmodel/explainabilityInput.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: It was not in the error message originally, but now it is in the case there is more than one tgw for the same couple of vpcs. I'll extract the transit gateway name from the vpc config in this case and put it in the error message which I'll reformat.

**Review by haim-kermany — DISMISSED** [review]

File: pkg/vpcmodel/explainabilityInput.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: done

**Review by haim-kermany — DISMISSED** [review]

File: pkg/vpcmodel/explainabilityInput.go [hunk 7]
```diff
@@ -107,19 +111,33 @@ func (configsMap MultipleVPCConfigs) getVPCConfigAndSrcDstNodes(src, dst string)
...
+		for cfgID, val := range configsWithSrcDstNodeMultiVpc {
+			return configsMap[cfgID], val.srcNodes, val.dstNodes, val.isSrcDstInternalIP, nil
+		}
+	// Single match of single vpc config, return it. Note that in this case there could be also matches in a
+	// multi vpc config that contains this vpc.
+	case len(configsWithSrcDstNodeSingleVpc) == 1:
+		for cfgID, val := range configsWithSrcDstNodeSingleVpc {
+			return configsMap[cfgID], val.srcNodes, val.dstNodes, val.isSrcDstInternalIP, nil
+		}
+	default: // len(configsWithSrcDstNodeSingleVpc) > 1 ||
```
Comment (adisos) [inline pkg/vpcmodel/explainabilityInput.go:137]: do not understand the logic here and the comments.. (such as "Note that in this case there could be also matches in a
multi vpc config that contains this vpc.")

**Review by adisos — CHANGES_REQUESTED** [review]
initial review, went over most files.. 
remaining: explainabilityPrint and explainabilityInput.

File: pkg/ibmvpc/examples/out/explain_out/tgwAnotherEnableDefaultDifFile_all_vpcs_explain.txt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+Connectivity explanation between vsi11-ky and vsi21a-ky within combined-vpc-local-tg-ky
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/examples/out/explain_out/tgwAnotherEnableDefaultDifFile_all_vpcs_explain.txt:1]: And indeed I've removed it in `c.VPC.AddressPrefixes()`
Look at the test files there after the change; if needed lets change them there

File: pkg/ibmvpc/vpc.go [hunk 9]
```diff
@@ -687,10 +699,27 @@ func (pgw *PublicGateway) AllowedConnectivity(src, dst vpcmodel.VPCResourceIntf)
...
 func (pgw *PublicGateway) AppliedFiltersKinds() map[string]bool {
 	return map[string]bool{vpcmodel.NaclLayer: true, vpcmodel.SecurityGroupLayer: true}
 }
 
+func (pgw *PublicGateway) StringPrefixDetails(src, dst vpcmodel.Node, verbose bool) (string, error) {
+	return "", nil
+}
+
+// a tgw prefix for explain
+type tgwPrefix struct {
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go:718]: done

File: pkg/ibmvpc/vpc.go [hunk 10]
```diff
@@ -755,6 +790,103 @@ func (tgw *TransitGateway) AllowedConnectivity(src, dst vpcmodel.VPCResourceIntf
...
+		return "", err
+	}
+	return fmt.Sprintf(" default prefix,  action: %s", actionName), nil
+}
+
+func (tgw *TransitGateway) tgwPrefixStr(prefix tgwPrefix) (string, error) {
+	// Array of prefix route filters for a transit gateway connection. This is order dependent with those first in the
+	// array being applied first, and those at the end of the array is applied last, or just before the default.
+	resStr := fmt.Sprintf("transit-connection: %s", *prefix.tc.Name)
+	if prefix.index == -1 { // default
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go:815]: done

File: pkg/vpcmodel/explainabilityInput.go (hunk 7 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go]: I improved the documentation of this function and of the following one. See if it is clearer now.
I also copied a bug fix from #502 in the following function.

File: pkg/ibmvpc/parser.go [hunk 11]
```diff
@@ -832,6 +833,12 @@ func getTgwObjects(c *datamodel.ResourcesContainerModel,
 			// TGW's destSubnets contains subnets from its connected VPCs which are contained within routes from its table
 			tgwMap[tgwUID].destSubnets = append(tgwMap[tgwUID].destSubnets, getVPCdestSubnetsByAdvertisedRoutes(tgwMap[tgwUID], vpc)...)
 			tgwMap[tgwUID].addSourceAndDestNodes()
+
+			// explainability related computation
+			// vpcApsPrefixes is a map from the vpc's ap to the index of the prefix that matches it, -1 for default if no match
+			// if getVPCAdvertisedRoutes completed without an error, so would getVpcApsPrefixes
+			vpcApsPrefixes, _ := getVpcApsPrefixes(tgwConn, vpc)
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/parser.go:840]: because, as mentioned here, there is no way `getVpcApsPrefixes` while `getVPCAdvertisedRoutes`  completed successfully. and unless `getVPCAdvertisedRoutes`  completed successfully we would not get here.

File: pkg/ibmvpc/tgw_analysis.go [hunk 12]
```diff
@@ -130,3 +130,41 @@ func prefixLeGeMatch(prefix *string, le, ge *int64, cidr string) (bool, error) {
...
+		match, err := prefixLeGeMatch(pf.Prefix, pf.Le, pf.Ge, cidr)
+		if err != nil {
+			return -1, err
+		}
+		if match {
+			return i, nil
+		}
+	}
+	// no match by pfList -- use default
+	return -1, nil
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go]: done

File: pkg/ibmvpc/vpc.go [hunk 13]
```diff
@@ -713,6 +742,12 @@ type TransitGateway struct {
 	destNodes   []vpcmodel.Node
 
 	region *Region
+
+	// maps APs of each VPC to the prefix that determines its connectivity status w.r.t. the tgw (allow/deny)
+	// if non default. Specifically, the map is from VPC UID to a map between the ap's ipBlock to the index of the matching prefix
+	// that determines its status;
+	// this struct can be though of as the "explain" parallel of availableRoutes; note that it also lists deny prefixes
+	vpcApsPrefixes map[string]map[*ipblock.IPBlock]tgwPrefix
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go]: Actually this map is not used as a map. Only the value of the `*ipblock.IPBlock` is used. 
Changing the map to used string and performing the required conversions seems to much work given that we don't actually need a map.
We can either add a comment regarding this issue, or change the struct used to avoid future confusion. In the latter case we will need to define a dedicated struct to that end, and use a slice of this newly defined struct. Any other idea? @adisos

[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go]: done

File: pkg/ibmvpc/vpc.go [hunk 14]
```diff
@@ -713,6 +742,12 @@ type TransitGateway struct {
 	destNodes   []vpcmodel.Node
 
 	region *Region
+
+	// maps APs of each VPC to the prefix that determines its connectivity status w.r.t. the tgw (allow/deny)
+	// if non default. Specifically, the map is from VPC UID to a map between the ap's ipBlock to the index of the matching prefix
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go]: Please look at the current documentation - hope its clear now

File: pkg/ibmvpc/tgw_analysis.go [hunk 15]
```diff
@@ -130,3 +130,41 @@ func prefixLeGeMatch(prefix *string, le, ge *int64, cidr string) (bool, error) {
...
+			return nil, err1
+		}
+		apIPBlock, err2 := ipblock.FromCidr(ap)
+		if err2 != nil {
+			return nil, err2
+		}
+		res[apIPBlock] = tgwPrefix{tc, prefixIndex}
+	}
+	return res, nil
+}
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go]: indeed. done.

File: pkg/ibmvpc/parser.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/parser.go]: anyways, after merging the functionality this comment is no longer relevant

File: pkg/ibmvpc/vpc.go [hunk 16]
```diff
@@ -755,6 +790,103 @@ func (tgw *TransitGateway) AllowedConnectivity(src, dst vpcmodel.VPCResourceIntf
...
+		return "allow", nil
+	}
+	return "deny", nil
+}
+
+func (tgw *TransitGateway) StringPrefixDetails(src, dst vpcmodel.Node, verbose bool) (string, error) {
+	prefix := tgw.prefixOfSrcDst(src, dst)
+	transitEnablesConn := vpcmodel.HasNode(tgw.sourceNodes, src) && vpcmodel.HasNode(tgw.destNodes, dst)
+	if verbose {
+		tgwRouterFilterDetails, err := tgw.tgwPrefixStr(*prefix)
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go:856]: This isn't a valid option.
If we got here then `func (tgw *TransitGateway) RouterDefined(src, dst vpcmodel.Node) bool` returned true in which case `prefix` isn't `nil` (may point to the default defined behavior)

File: pkg/ibmvpc/vpc.go [hunk 17]
```diff
@@ -755,6 +790,103 @@ func (tgw *TransitGateway) AllowedConnectivity(src, dst vpcmodel.VPCResourceIntf
...
+func actionNameStr(action *string) (string, error) {
+	actionBool, err := parseActionString(action)
+	if err != nil {
+		return "", err
+	}
+	if actionBool {
+		return "allow", nil
+	}
+	return "deny", nil
+}
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go]: done. note that it effects the printing - "permit" instead of "allow" which is non-consistent with the acl and sg wording in the same printing.
Perhaps we should consider making it consistent by changing permit to allow?

File: pkg/ibmvpc/vpc.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/vpc.go]: the name is still the same.. better rename

[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/vpc.go]: I think the documentation for this field it still outdated? also can it be made more simple/concise ?

[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/vpc.go]: possible changes:
```
	// maps each VPC UID to a slice - listing for each of its AP (as IPBlock) details of the matching filter
       // Specifically, these details include the specific transit connection and the index of the matching
	// filter if exists (index "-1" is for default )
	// this struct can be thought of as the "explain" parallel of availableRoutes; note that unlike availableRoutes it also lists deny prefixes
```

[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go]: No, its not supposed to be. Here, this is what you should have seen (was already pushed). Is this what you referred to?
```

	// maps each VPC to a struct listing for each of its address prefix details of the filter prefix determining its behavior w.r.t.
	// the tgw. Specifically, these details include the specific transit connection and the index of the matching
	// (allow/deny) filter prefix rule if exists, or -1 if the behavior is determined by the specified default of the transit connection
	// this struct can be though of as the "explain" parallel of availableRoutes; note that unlike availableRoutes it also lists deny prefixes
	vpcsAPToFilters map[string][]IPBlockPrefixFilter
```

**Review by adisos — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/tgw_analysis.go [hunk 18]
```diff
@@ -75,19 +81,21 @@ func isCIDRMatchedByPrefixFilters(cidr string, tc *datamodel.TransitConnection)
 	pfDefault := tc.PrefixFiltersDefault
 
 	// TODO: currently assuming subnet cidrs are the advertised routes to the TGW, matched against the prefix filters
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go:83]: Why is this todo is to be removed?

File: pkg/ibmvpc/vpc.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/vpc.go]: I'd keep it , as these are the terms the user is familiar with..
@zivnevo @kyorav  any other thoughts here?

File: pkg/ibmvpc/tgw_analysis.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/tgw_analysis.go:83]: it's my old comment that should have been removed.. no longer assuming subnet cidrs are the advertised routes to the TGW, using address prefixes instead.

[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go:83]: right

File: pkg/ibmvpc/vpc.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (kyorav) [reply pkg/ibmvpc/vpc.go]: I agree with Adi, it is best to be consistent with the terms used by IBM Cloud

File: pkg/ibmvpc/tgw_analysis.go [hunk 19]
```diff
@@ -48,25 +48,31 @@ func validateAddressPrefixesExist(vpc *VPC) {
...
 // thus advertised to a TGW
-func getVPCAdvertisedRoutes(tc *datamodel.TransitConnection, vpc *VPC) (res []*ipblock.IPBlock, err error) {
+func getVPCAdvertisedRoutes(tc *datamodel.TransitConnection, vpc *VPC) (advertisedRoutesRes []*ipblock.IPBlock,
+	vpcApsPrefixesRes []IPBlockPrefixFilter, err error) {
 	validateAddressPrefixesExist(vpc)
-	for _, ap := range vpc.addressPrefixes {
-		matched, err := isCIDRMatchedByPrefixFilters(ap, tc)
+	vpcApsPrefixesRes = make([]IPBlockPrefixFilter, len(vpc.addressPrefixes))
+	for i, ap := range vpc.addressPrefixes {
+		prefixIndex, matched, err := getCIDRMatchedByPrefixFilters(ap, tc)
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go:57]: done in all relevant places

File: pkg/ibmvpc/tgw_analysis.go [hunk 20]
```diff
@@ -75,19 +81,21 @@ func isCIDRMatchedByPrefixFilters(cidr string, tc *datamodel.TransitConnection)
...
-			return false, err
+	for prefixIndex, pf := range pfList {
+		match, err1 := prefixLeGeMatch(pf.Prefix, pf.Le, pf.Ge, cidr)
+		if err1 != nil {
+			return minusOne, false, err1
 		}
 		if match {
-			return parseActionString(pf.Action)
+			action, err = parseActionString(pf.Action)
+			return prefixIndex, action, err
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/tgw_analysis.go]: right, confused..

File: pkg/ibmvpc/vpc.go [hunk 21]
```diff
@@ -6,13 +6,16 @@ import (
...
 	"github.com/np-guard/models/pkg/connection"
 	"github.com/np-guard/models/pkg/ipblock"
-
 	"github.com/np-guard/vpc-network-config-analyzer/pkg/vpcmodel"
 )
 
-// /////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////
+
+const minusOne = -1
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go:17]: done

File: pkg/ibmvpc/tgw_analysis.go [hunk 22]
```diff
@@ -48,25 +48,31 @@ func validateAddressPrefixesExist(vpc *VPC) {
...
+		vpcApsPrefixesRes[i] = IPBlockPrefixFilter{apIPBlock, tgwPrefixFilter{tc, prefixIndex}}
 	}
-	return res, nil
+	return advertisedRoutesRes, vpcApsPrefixesRes, nil
 }
 
-func isCIDRMatchedByPrefixFilters(cidr string, tc *datamodel.TransitConnection) (bool, error) {
+// gets for a given address prefix the matching prefix filter index and its action (allow = true/deny = false)
+// if there is no specific prefix filter then gets the default defined behavior
+func getCIDRMatchedByPrefixFilters(cidr string, tc *datamodel.TransitConnection) (prefixIndex int, action bool, err error) {
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go]: done

[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go]: done the first time of this comment

File: pkg/ibmvpc/tgw_analysis.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/tgw_analysis.go]: `matched` was not renamed to `isPermitAction`

File: pkg/vpcmodel/explainabilityPrint.go [hunk 23]
```diff
@@ -118,79 +188,66 @@ func noConnectionHeader(src, dst string, connQuery *connection.Set) string {
...
 	}
-	resStr += filtersEffectStr + "\n" + rulesStr
-	return resStr
+	resComponents = append(resComponents, path, details)
+	return strings.Join(resComponents, newLine)
 }
 
 // returns a string with a summary of each filter (table) effect; e.g.
 // "Egress: security group sg1-ky allows connection; network ACL acl1-ky blocks connection
 // Ingress: network ACL acl3-ky allows connection; security group sg1-ky allows connection"
```
Comment (adisos) [inline pkg/vpcmodel/explainabilityPrint.go:214]: function returns two strings - update documentation

File: pkg/vpcmodel/explainabilityPrint.go [hunk 24]
```diff
@@ -118,79 +188,66 @@ func noConnectionHeader(src, dst string, connQuery *connection.Set) string {
...
+		ingressRulesHeader = "Ingress: " + ingressRulesHeader + newLine
 	}
-	if egressRulesStr != "" && ingressRulesStr != "" {
-		return egressRulesStr + "\n" + ingressRulesStr
-	}
-	return egressRulesStr + ingressRulesStr
+	return egressRulesHeader, ingressRulesHeader
 }
 
 // returns a string with a detailed list of relevant rules; e.g.
```
Comment (adisos) [inline pkg/vpcmodel/explainabilityPrint.go:232]: same here

File: pkg/ibmvpc/tgw_analysis.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/ibmvpc/tgw_analysis.go]: renamed now

File: pkg/vpcmodel/explainabilityPrint.go (hunk 23 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityPrint.go:214]: fixed

File: pkg/vpcmodel/explainabilityPrint.go (hunk 24 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityPrint.go:232]: also here

**Review by adisos — APPROVED** [review]
LGTM
